src/tiledb/cloud/utilities/consolidate.py

```diff
@@ -0,0 +1,248 @@
+from collections import defaultdict
+from os.path import basename
+from typing import Any, Mapping, Optional, Sequence
+
+import tiledb
+from tiledb.cloud import dag
+from tiledb.cloud.rest_api.models import RetryStrategy
+from tiledb.cloud.utilities import get_logger
+from tiledb.cloud.utilities import max_memory_usage
+from tiledb.cloud.utilities import run_dag
+
+MAX_FRAGMENT_SIZE = 1 << 30  # bytes
+
+
+def group_fragments(
+    array_uri: str,
+    *,
+    config: Optional[Mapping[str, Any]] = None,
+    group_first_dim: bool = True,
+) -> Sequence[Sequence[tiledb.FragmentInfo]]:
+    """
+    Get a list of fragment info objects, optionally grouping fragments that have the
+    same value for the first dimension.
+
+    :param array_uri: array URI
+    :param config: config dictionary, defaults to None
+    :param group_first_dim: group by first dimension, defaults to True
+    :return: list of lists of fragment info objects
+    """
+
+    logger = get_logger()
+    logger.info("Grouping fragments for array %r", array_uri)
+
+    with tiledb.scope_ctx(config):
+        fragment_info_lists = defaultdict(list)
+
+        fis = tiledb.FragmentInfoList(array_uri)
+
+        for fi in fis:
+            if group_first_dim:
+                fragment_info_lists[fi.nonempty_domain[0]].append(fi)
+            else:
+                fragment_info_lists["all"].append(fi)
+
+        # Create a list of lists of fragment info objects.
+        results = [x for x in list(fragment_info_lists.values())]
+
+        logger.info(f"{len(fis)} fragments grouped into {len(results)} groups")
+        logger.info(f"max memory usage: {max_memory_usage() / (1 << 30):.3f} GiB")
```

Comment:
minor: rather than substituting things in on the logger string itself, instead use `%` placeholders (so that if the logging event does not happen, the string doesn't need to get constructed):

```python
logger.info("%d fragments grouped into %d groups", len(fis), len(results))
```